HPCC-28059 Validate requests before searching WUs and queries

wangkx opened the pull request:
The code is added to make sure that search requests do not have
invalid character as xpath values. The changes in this PR focus
on the following ESP services:
1. WsWorkunits.WUQuery
2. WsWorkunits.WULightWeightQuery
3. WsWorkunits.WUListQueries
4. WsWorkunits.WUQueryDetails
5. WsWorkunits.WUQueryDetailsLightWeight
6. FileSpray.GetDFUWorkunits

The code is also added to validate the workunit ID in
FileSpray.GetDFUWorkunit request.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

wangkx committed:
HPCC-28059 Validate requests before searching WUs and queries

The code is added to make sure that search requests do not have
invalid character as xpath values. The changes in this PR focus
on the following ESP services:
1. WsWorkunits.WUQuery
2. WsWorkunits.WULightWeightQuery
3. WsWorkunits.WUListQueries
4. WsWorkunits.WUQueryDetails
5. WsWorkunits.WUQueryDetailsLightWeight
6. FileSpray.GetDFUWorkunits

The code is also added to validate the workunit ID in
FileSpray.GetDFUWorkunit request.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented:
@afishbeck @ghalliday please review, especially on: 1. the contents of the validXPathValueChrs; 2. any other validatation should be added (in this PR or in future's PR).

afishbeck commented on esp/services/ws_fs/ws_fsService.cpp:
```diff
@@ -3677,3 +3680,19 @@ bool CFileSprayEx::onGetDFUServerQueues(IEspContext &context, IEspGetDFUServerQu
+        throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "GetDFUWorkunits: Invalid StateReq %s.", req.getStateReq());
+    if (!isEmptyString(req.getJobname()) && !isValidXPathValue(req.getJobname()))
+        throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "GetDFUWorkunits: Invalid Jobname %s.", req.getJobname());
+    if (!isEmptyString(req.getPublisherWuid()) && !isValidXPathValue(req.getPublisherWuid()))
```
I think for PublisherWuid you can do  !looksLikeAWuid(req.getPublisherWuid(), 'P').

wangkx replied:
I just made the change.

wangkx committed:
HPCC-28059 Revise based on review

Check the PublisherWuid using looksLikeAWuid().
Remove the varialbe p from isValidXPathValue().

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

afishbeck commented on system/jlib/jptree.hpp:
```diff
@@ -313,6 +313,22 @@ inline static bool isValidXPathChr(char c)
     return ('\0' != c && (isalnum(c) || strchr(validChrs, c)));
 }
 
+static const char *validXPathValueChrs = "!#$%^&()+,-.:;?@_{|}~"; //Do I miss any?
```
Hmmm.  The level of validation could and probably should be a lot stricter.  For each request parameter we should first ask a few questions:
1. Is it ever stored in DALI/PTree as an element?  If so the allowed characters are a lot less.
2. Is it stored in Ptree as a string?
3. Can it be treated as a restricted name type... with a limited range of characters.
4. Is the parameter allowed to have wild cards?

For checking xpath elements rather than value strings there are existing methods in ptree.cpp
inline static bool isValidXPathStartChr(char c)
{
    return ('\0' != c && (isalpha(c) || strchr(validStartChrs, c)));
}

static const char *validChrs = ":_.-";
inline static bool isValidXPathChr(char c)
{
    return ('\0' != c && (isalnum(c) || strchr(validChrs, c)));
}

For a lot of parameters that are names... cluster names, etc.  We can be very strict.  Perhaps only allowing alphanum and '.' and '-' etc (what others?)?

afishbeck replied:
If it is stored as an element we might not even allow ':' which would act like a namespace... although ptree could still handle it so it may be ok depending on what it is.

afishbeck replied:
You could have at least 3 functions:

isValidXMLElementName(char *s);
isValidXPathValue(char *s);
isValidName(char *s, const char *extraAllowedCharacters);

example:
isValidName("my-secret.101", "-.");  //returns true
isValidName("my-secret.101[1]", "-.");  //returns false

wangkx replied:
Good comments. I thought about to create this PR as a draft to collect comments.
I name it as validXPathValueChrs and, in this PR, it is only used to verify an xpath value: xpath='**value**'. I haven't thought about validating elements and names yet.

wangkx replied:
All of the functions can have the default allowed characters and extraAllowedCharacters. The challenge is which characters should be in the default allowed.

afishbeck commented on system/jlib/jptree.hpp:
```diff
@@ -313,6 +313,22 @@ inline static bool isValidXPathChr(char c)
+}
+inline static bool isValidXPathValue(const char *v)
+{
+    const char *p = v;
```
I don't think you actually need p.  The following should work:
while (*v)
{
     if (!isValidXPathValueChr(*v++)) return false;
}

wangkx replied:
Yes, I do not need p.
The do...while is used to check an empty v.

wangkx committed:
HPCC-28059 Revise based on review

Add isValidXMLElementName() and isValidName()

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented:
@ghalliday please review.

ghalliday commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -4908,4 +4908,56 @@ void CWsWuEmailHelper::send(const char* body, const void* attachment, size32_t l
+
+void CWsWuValidateHelper::validateWUListQueriesRequest(IEspWUListQueriesRequest& req)
+{
+    if (!isEmptyString(req.getQuerySetName()) && !isValidXPathValue(req.getQuerySetName()))
```
One further follow on comment.  It would simplify the code (and make it easier to read) if each isvalidX() function had a bool required.  E.g.

```
bool isValidQuerySetName(const char * name, bool required)
{
   if (!name || *name == '\0')
      return !required;
...
}

```
Then the call will be 

`if (isValidQuerySetName(req.querySetName(), false))
`
which is much easier to read and check.

wangkx replied:
Should we add another bool allowWild?

ghalliday reviewed (commented):
@wangkx A more significant comment than I expected to make.  Feel free to discuss.

wangkx commented:
Thanks for the comments which suggest the checks from another perspective: logical entities. Different logical entities may have different restrictions. Creating an isValid...() for each logical entity may be better for checking those restrictions. Of course, it may take time to design/create those isValid...().

wangkx commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -4908,4 +4908,56 @@ void CWsWuEmailHelper::send(const char* body, const void* attachment, size32_t l
+    validateWUQueryRequestCommon("WUQuery", req);
+    if (!isEmptyString(req.getJobname()) && !isValidXPathValue(req.getJobname()))
+        throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "WUQuery: Invalid Jobname %s.", req.getJobname());
+    if (!isEmptyString(req.getECL()) && !isValidXPathValue(req.getECL()))
```
```searching for the text "\ should produce a filter on \"\\```

Is the issue reported in the JIRA caused by incorrect encoding? One exception there is: 
XPath Exception: Qualifier brace unclosed\nin xpath = [@jobName=?\"*\"AAA*\"]
Could the extra `\"` cause the exception (the ] is not inside quotes)?

wangkx commented:
Redesign is needed for this PR. I am closing this PR for now.
